### ISSUE TITLE
Using angular-phoenix within a controller

### ISSUE THREAD

**Issue by declanelcocks** [MikaAK/angular-phoenix#1]
Hey, first of all it's great to see a Phoenix Angular project. I'm currently trying to move over my Phoenix socket to my Angular app using your project and having some difficulties. It would be great to see a simple demo app for your framework, but I'll ask my question anyway. 

I'm just trying to implement it on a single page of my app, so ideally I'd just like to connect to the socket and interact with the socket from within the controller for that page. Could you give some help or examples for using angular-phoenix from within a controller? 

I tried using the phoenix.js file as you said, but also ran into errors there saying that phoenix.js was not private or something like that.

**Comment by declanelcocks** [comment]
I have setup Phoenix in my controller using:

`.controller('ChatController', ['$scope', 'Phoenix',
    function($scope, Phoenix) {
        // Scope variables
    }])`

But when I add your `Phoenix.join` method I get an error for `Unexpected token: =>`. If I just try using the very basic `Phoenix.join` with something like this:

`Phoenix.join("rooms:lobby", {});`

Then I get this error:

`Cannot read property '$$hashKey' of undefined`

**Comment by MikaAK** [comment]
Unfortunately my library is a tad outdated. I haven't had the chance to upgrade this from `0.11.0` to `0.13.0` which has some new phoenix.js changes! I'll try and get around by updating this by friday! 

What version of Phoenix are you running?
`Unexpected token: =>.` sounds like your trying to run es6 code in the browser are you using the code from `dist` or `src`? Also are you using the built version of phoenix.js or the es6 one! 

I've added a few more examples to the readme as well!

**Comment by declanelcocks** [comment]
I'm using the latest version of Phoenix at the minute, and I'm using the `dist` version of angular-phoenix. For `phoenix.js` I'm using the file in the `priv` folder from their download, I tried using the file from `web -> static` which says in the comments it's the plain JS version but this gave a whole load of new errors. Can't seem to remove this `=>` error anyway, here's my code from my controller so far, but I imagine I am missing something looking at your example.

`(function() {

```
var app = angular.module('timeline')

.controller('ChatController', ['$scope', 'Phoenix',
function($scope, Phoenix) {
    // Scope variables

    Phoenix.join('rooms:lobby', {}).promise
      .then(chann => {
        // Now our callbacks will get removed on scope destruction
        chann.on(scope, 'message', handler)
        chann.on('message', hander)
      });

}])
```

}());`

And in my `app.js` file I have the `app.config` settings listed on the example.

**Comment by MikaAK** [comment]
`=>` is an [es6 syntax](https://github.com/lukehoban/es6features#arrows) if you aren't useing es6 you will have to use `function(chann) { //dostuff}`

I'm not sure this will work with the newest version of phoenix yet but I could be wrong!

**Comment by declanelcocks** [comment]
Ah right, that took care of that error. Sorry I haven't really taken too much of a look at ES6 yet. After resolving that I am just back to my original error of `Cannot read property '$$hashKey' of undefined`, specifically here:

`at joinChannel (http://localhost:8888/wechat_frontend/js/angular-phoenix.js:119:22)`
`at Object.join (http://localhost:8888/wechat_frontend/js/angular-phoenix.js:154:17)`

Thanks for your help so far by the way

**Comment by MikaAK** [comment]
Ah, I was using `Promise` before which if you're not in es6 mode doesn't exist. I've switched it to `$q`. Run an update and let me know how that goes.

**Comment by declanelcocks** [comment]
Thanks, I'm just getting this error now:

`TypeError: socket.join is not a function
    at $get.joinRes (http://localhost:8888/wechat_frontend/js/angular-phoenix.js:102:26)`

which is this line in `angular-phoenix.js`:

`joinRes = function (resolve, reject) {
        channel = socket.join(name, message);`

I believe the `phoenix.js` file is included with no issues (I am using the file from `priv > static` from the phoenix Git repo), and my variable names in the controller are ok as it is successfully using `angular-phoenix.js`. Can't see the issue just yet. Hopefully if I can get this up and running I'll be doing quite a lot with it for a project I'm working on, so I'd be happy to contribute some example use cases to your GitHub repo in the future.

**Comment by MikaAK** [comment]
This seems to be due to the recent changes with the front end code. I used the [Phoenix.js](https://raw.githubusercontent.com/phoenixframework/phoenix/918afc1a4f2d1c40154f156e101acf668b5ec93e/priv/static/phoenix.js) from 0.11.0 while it has been updated to [0.13.0](https://github.com/phoenixframework/phoenix/blob/master/CHANGELOG.md) and has a new api for the client.

**Comment by declanelcocks** [comment]
Ah right, that fixed that. I'll let you know how I get on with furthering the app, and if I have time this week I'll try to update it to the new API then. I'll need to move onto listening for new messages next so I'll see how that goes, I assume most of the API is the same as using normal Phoenix!

**Comment by declanelcocks** [comment]
Here's the code using Phoenix without AngularJS if you are interested, I'm trying to convert it over to Angular. I've pretty much just used the exact same code to connect and listen for new messages which is working, so that's great to see :)

```
let socket = new Socket("ws-address-here")
socket.connect()
socket.join("groups:1", {"token": "abc"}).receive("ok", function(chan) {
 console.log("Welcome to Chat!")

 chan.on("new_msg", payload => {
   messagesContainer.append(`<br/>[${Date()}] ${payload.data.user_info.nickname} says: ${payload.data.content}`)
 })

 chatInput.off("keypress").on("keypress", function(event) {
   if(event.keyCode === 13){
     chan.push("new_msg", {data: 
       {
         content: chatInput.val(),
         kind: 'text',
         group_id: 1,
         user_info: {
           user_id: 1,
           nickname: 'test',
           avatar_url: 'http://abc.com/abc.png'
         }
       }
     })
     chatInput.val("")
   }
 })

})
```

**Comment by MikaAK** [comment]
The only major difference is don't forget to pass `scope` to `chan.on` so the callback is destroyed when the scope is. I've updated the codebase to use the latest phoenix however it no longer seems to function for me. I won't be able to test this out fully due to time until I can upgrade my server to 0.13. Due to this I released the code under a candidate as I'll need testers!

**Comment by declanelcocks** [comment]
You mean in the existing code above, if I implement the code below, then the callback will be destroyed when the scope is? I don't seem to be able to destroy the connection when leaving the page at the moment, but I think I'm not writing it correctly.

That would be great if the code was updated to work with `0.13.0` so let me know if I can help with that. The backend of the app I'm using is actually done using the latest version, but I'm using version `0.11.0` of `phoenix` and `angular-phoenix` which is surprisingly working.

```
Phoenix.socket.connect();

  Phoenix.join("name", {"token": "abc"}).receive("ok", function(chan) {
    console.log("welcome to chat");

    chan.on($scope, "new_msg", handler);

  });
```

**Comment by MikaAK** [comment]
If you want to try out version `0.3.0-rc.1` that should work for `0.13.0` but It could need work. 

For the second one, I actually never added logic to disconnect when the scope is destroyed. I just added it and fixed the scope destruction for `chan.on`

This new added feature is untested but I shall do so this evening. 
I put this under `0.2.5-rc.1` since I can't test it yet! 

It allows you to do something like this. 

``` javascript
Phoenix.join($scope, "name", {"Token": "ABC"}) // Will leave channel when scope is destroyed. 
```

Thanks for dealing with all the bugs and helping me test this out!

**Comment by declanelcocks** [comment]
Yeah, I'm using the [0.2.4 branch](https://github.com/MikaAK/angular-phoenix/tree/0.2.4-phoenix-0.11.0/dist) to test out the scope destruction. I've added in a destroy function to my controller like so:

```
destroy = function() {
  console.log('scope destroyed');
}

// add listener for destroy
$scope.$on('$destroy', destroy);
```

The scope is indeed being destroyed as the message is shown in console when moving to another page. But the channel is still connected as when I go back to the page I still get the `chan.on is not a function` error. I tried manually implementing `Phoenix.leave("name");` but this didn't seem to work. I'll take a look at the `0.2.4` code to see why it isn't working.

As for the other issue, when I try using `0.3.0` with `0.13.0` I receive `TypeError: Phoenix.join(...).receive is not a function`. I'll continue to work with `0.2.4` and `0.11.0` for now to see if I can get the scope destruction working.

**Comment by declanelcocks** [comment]
I have managed to get it working successfully by doing this:

```
Phoenix.join($scope, "name", {"token": "abc"}).receive("ok", function(chan) {
    // Do some events here
    $scope.$on('$destroy', function () {
        return chan.leave();
    });
});
```

This works so that I can leave the page, go back to the page and start chatting again. Doing this though I don't actually have to pass it the `$scope` when joining a channel. I see that in `angular-phoenix` some code similar to this already exists to do this, but I can't get it to run yet. It would definitely be better to automatically leave the channel on scope destruction as you said before.

**Comment by MikaAK** [comment]
I've fixed the bug with allowing you to pass it into the `on` event via `on(scope, eventName, callback)` as well as taken a look into `join(scope, eventName, message)`. If a scope is passed to join when the scope is destroyed the channel will be left left! If that's not working for you make sure your scope is being destroyed!

**Comment by declanelcocks** [comment]
I'll give it a try when I get to my laptop and let you know!

**Comment by declanelcocks** [comment]
I still seem to get the same issue with 0.2.5, it doesn't re-join the channel when I go back to the chat page. Pasted the controller [here](http://jsbin.com/zocowolone/1/edit?js), maybe I'm setting it up slightly different to how you are expecting it to be implemented in Angular.

**Comment by MikaAK** [comment]
That looks like it should work. It's hard for me to say what's wrong but best I can tell is for some reason 
[these lines](https://github.com/MikaAK/angular-phoenix/blob/phoenix-0.11.0/dist/angular-phoenix.js#L163-L167) aren't properly working! If you try putting a `debugger` inside the destroy can you confirm that it's getting called that may help to debug the issue

**Comment by declanelcocks** [comment]
Yeah, the function is not getting called. When using the commented out code, the scope is definitely being destroyed when changing page.

`resChannel.promise.then` outputs the following to console:

```
function (a,b,c){var d=new g;this.$$state.pending=this.$$state.pending||[];this.$$state.pending.push([d,a,b,c]);0<this.$$state.status&&
f(this.$$state);return d.promise}
```

But it never actually goes into the function when changing page.

**Comment by MikaAK** [comment]
The only thing left I can think of is check [here](https://github.com/MikaAK/angular-phoenix/blob/phoenix-0.11.0/dist/angular-phoenix.js#L106-L108). If you breakpoint the code here what is the value of `chan`?

**Comment by declanelcocks** [comment]
If I insert a log to output `chan` here, I don't get any output when I load the page or when I navigate away from the page. I can output `channel` to console no problem just before that, but inside the `channel.after` section of code I don't get any output.

**Comment by MikaAK** [comment]
At this point I must say I'm at a loss. For me chan is a instance of `Phoenix.Channel`. I'm guessing this is due to a versioning thing as if you see [here](https://gist.github.com/chrismccord/3603fd2735019f86c74b#file-gistfile1-md), the method of getting the channel is different [the new api](https://gist.github.com/chrismccord/0a3bf5229801d61f219b)

I'm just currently updating my server so I'll updated this client as soon as I can!

**Comment by declanelcocks** [comment]
Hm, I'll look closer at the changes. Seems strange that it should work for you and not for me, I'm using `phoenix 0.11.0` and `angular-phoenix 0.2.5`. For now I'll just leave the manual leave function in on scope destruction to move onto some other features and see if I encounter any other issues.

**Comment by declanelcocks** [comment]
Just to clarify, if I add in a `console.log` here then it does output an instance of `Phoenix.Channel`:

```
joinRes = function (resolve, reject) {
  channel = socket.join(name, message);

  channels.set(name, { status: 'fetching', channel: channel });

  console.log(channel);

  channel.after(5000, reject).receive(function (chan) {
    return resolve(chan);
  });
};
```

But if I put any log inside `channel.after`, then I get no output to the console.

**Comment by MikaAK** [comment]
I just got a chance to update to the newest version of Phoenix. I've tested and gotten auto leaving on scope destruction to work in the newest version. Let me know if everything works for you now!

**Comment by MikaAK** [comment]
I've just released the newest version which uses the format from Phoenix 0.13. Give that a try and let me know! I've tested this and it's all working for me!

**Comment by declanelcocks** [comment]
What syntax are you using for connecting? Before I was using `Phoenix.socket.connect()` in my controller for this page to connect.

**Comment by declanelcocks** [comment]
How are you making the initial connection to the socket in the controller? Before I was just using `Phoenix.socket.connect()` and then calling `Phoenix.join(name)`.

You can check how I'm implementing it [here](http://jsbin.com/qupamatuxe/1/edit?js), something isn't quite right with it I think.

**Comment by MikaAK** [comment]
You do not need to call socket.connect manually. If you see in the readme unless you specifically set `autoJoin` to false it will automatically call `socket.connect()`.

**Comment by MikaAK** [comment]
Check [here](http://www.phoenixframework.org/v0.13.1/docs/channels) for documentation regarding how to use the client side. You need to pass the message to `.chan` not `.join`. You can also bind your `.on` event's outside the `.receive`. I forgot to update the readme till just now but essentially, asides for the extra scope options and not having to worry about the socket connecting, the documentation from Phoenix for the client applies to this.

**Comment by declanelcocks** [comment]
I see, I'm following that and using `Phoenix.connect()` to connect to the socket in the controller. I'm not sure if this is a Phoenix bug right now though as I'm getting `failed: WebSocket is closed before the connection is established.` when connecting from [this] (https://github.com/phoenixframework/phoenix/blob/master/priv/static/phoenix.js#L603) part of `phoenix.js`.

**Comment by declanelcocks** [comment]
Thanks, I'll take a look. I've followed the example in the Phoenix docs by using `Phoenix.connect()` to connect to the websocket, but I still get no output to confirm I am connected so I'll keep checking, must be something I have missed.

**Comment by declanelcocks** [comment]
Thanks, I've got it working no problem. I spent a few hours playing around with an error where I couldn't pick up the messages sent from the client, and also couldnt show the welcome message when joining a channel. Turns out this was because of an update [here](https://github.com/phoenixframework/phoenix/commit/7afda674ff62e0605bbfd2506b1afe1255ae1f63#diff-9575e448d78dc168ca82587a2d81a700) to the `phoenix.js` file. I replaced the file with the one that you linked on `angular-phoenix` and it is working great now with scope destruction causing it to close the connection too, so thanks for your work.

From the new update, there are quite a few changes that effect it so I'll take a look.
